lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Pat.hs

```diff
@@ -0,0 +1,104 @@
+{-# LANGUAGE LambdaCase #-}
+
+module LambdaBuffers.Compiler.TypeClass.Pat (
+  Pat (..),
+  toProd,
+  toRec,
+  toSum,
+  patList,
+  matches,
+) where
+
+import Data.Text (Text)
+
+{- A simple ADT to represent patterns.
+
+Note that this ADT allows us to represent nonsensical types (i.e. we can "put the wrong pattern in a hole").
+This could be ameliorated by using a GADT, which would give us correct-by-construction patterns at the
+cost of significantly more complex type signatures.
+-}
+
+data Pat
```

Comment:
I have a little beef with Pat that can be summed up below:

1. `VarP` I feel is out of place, consider the formulation of what a unification term is https://hackage.haskell.org/package/unification-fd-0.10.0.1/docs/src/Control-Unification-Types.html#UTerm:

```haskell
data UTerm t v = UVar v | UTerm t (UTerm t v)
```

A unification variable is treated explicitly and specially by the underlying unification machinery. In a sense you're doing the same with `VarP`, but it's not obvious to the reader what the 'roles' actually are and that can get confusing when we're doing this on `Ty` which itself has a concept of 'variable'.

Note that in the below formulation VarP is not there:

```haskell
data Pat a
  = 
    Name Text
  | ModuleName [Text] -- also stupid, also necessary -_-
  | Opaque
  | Nil -- Nil and :*  are hacks to write rules for ProdP and SumP. A bare Nil == Unit
  | a :* a -- cons
  | a := a 
  | RecP a 
  | ProdP a {- Pat arg should be a list of "Bare types" -}
  | SumP a {- where the Pat arg is expected to be (Constr l t :* rest) or Nil, where
                rest is either Nil or a tyList of Constrs -}
  | RefP a a {- 1st arg should be a ModuleName  -}
  | AppP a a {- Pattern for Type applications -}
  | DecP a a a
  deriving stock (Show, Eq, Ord, Functor, Foldable, Traversable)

infixr 5 :*

instance Unifiable Pat where
  zipMatch :: Pat a -> Pat a -> Maybe (Pat (Either a (a, a)))
  zipMatch (Name l) (Name r) = if l == r then Just $ Name l else Nothing
  zipMatch (ModuleName l) (ModuleName r) = if l == r then Just $ ModuleName l else Nothing
  zipMatch Opaque Opaque = Just Opaque
  zipMatch Nil Nil = Just Nil
  zipMatch (l :* r) (l' :* r') = Just (Right (l, l') :* Right (r, r'))
  zipMatch (l := r) (l' := r') = Just (Right (l, l') := Right (r, r'))
  zipMatch (RecP l) (RecP r) = Just (RecP $ Right (l, r))
  zipMatch (ProdP l) (ProdP r) = Just (ProdP $ Right (l, r))
  zipMatch (SumP l) (SumP r) = Just (SumP $ Right (l, r))
  zipMatch (RefP mn n) (RefP mn' n') = Just (RefP (Right (mn, mn')) (Right (n, n')))
  zipMatch (AppP f a) (AppP f' a') = Just (AppP (Right (f, f')) (Right (a, a')))
  zipMatch (DecP tn mn b) (DecP tn' mn' b') = Just (DecP (Right (tn, tn')) (Right (mn, mn')) (Right (b, b')))
  zipMatch _ _ = Nothing
```

2. Obscured ground terms

A logic program must be able to convey the following fundamental roles:
1. What are the 'ground terms',
2. How are 'ground terms' mapped to 'unifiable terms' (that unification machinery can handle)
3. What's a unification variable.

Ground terms are your domain terms sort of, simply defined as terms without variables. For us here, that just means ProtoCompat.Types types. For example:

```haskell
data Ty
  = TyVarI TyVar
  | TyAppI TyApp
  | TyRefI TyRef
  deriving stock (Show, Eq, Ord, Generic)
```

In Prolog this would correspond to the following ground terms (values of our ProtoCompat.Types):

```prolog
example(ty_var(a)).
example(ty_app(ty_local_ref(maybe), ty_local_ref(int))).
example(ty_ref(int)).
```
And in Prolog you magically get the ability to unify over 'any and all' ground terms:

```prolog
?- example(ty_app(ty_local_ref(TyName), R)).
R = ty_local_ref(int)
TyName = maybe
```

In Haskell, we need to be explicit about that magic. That's essentially the beef I have with this ad-hoc (not incorrect!) formulation of Pat.
Ideally, I'd see Unifiable instances defined on ProtoCompat.Types in some way, rather than being in the current situation:
1. https://github.com/mlabs-haskell/lambda-buffers/blob/b36982041bde869f0b6b27d7f98711352521d994/lambda-buffers-compiler/src/LambdaBuffers/Compiler/KindCheck/Variable.hs#L8
2. https://github.com/mlabs-haskell/lambda-buffers/blob/b36982041bde869f0b6b27d7f98711352521d994/lambda-buffers-compiler/src/LambdaBuffers/Compiler/KindCheck/Inference.hs#L109
3. Pat - even though correct, it's just introducing yet another mapping of our ProtoCompat.Types in an indirect way.

Reply:
See my slack thread. We're looking at this in two different ways: You're coming at this from a "this is a logic program" perspective and I'm coming at this from a "this is pattern matching" perspective. 

We probably could rewrite everything the way you want, but I suspect that doing so wouldn't really give us any tangible benefits. From the perspective of pattern matching `VarP` and `Pat` in general are _fine_. More than that, they're necessary: You need some notion of a pattern on which to do the matching and you need variable patterns to do any kind of useful matching.

Reply:
I'm going to click 'resolve' because I've explained (on slack & in #37)why what I'm doing here is fundamentally distinct from unification / logic programming. The three issues you raised are no longer relevant in light of the realization that pattern matching and unification are qualitatively different.